# Assignment: Data Wrangling

### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** [To be provided]

**Date:** [To be provided]

### Question 1 Part 1

In [8]:
!pip install pandas

In [11]:
import pandas as pd
data = pd.read_csv("data/airbnb_hw.csv")
print("First 20 values of the 'Price' column:")
print(data["Price"].head(20))

print("The Data Type of the 'Price' column:")
print(data["Price"].dtype)

print(f"Missing values: {data['Price'].isna().sum()}")

print("Value counts of the 'Price' column (top 20):")
data["Price"].value_counts().tail(20)

First 20 values of the 'Price' column:
0     145
1      37
2      28
3     199
4     549
5     149
6     250
7      90
8     270
9     290
10    170
11     59
12     49
13     68
14    285
15     75
16    145
17    100
18    150
19    700
Name: Price, dtype: str
The Data Type of the 'Price' column:
str
Missing values: 0
Value counts of the 'Price' column (top 20):


Price
509       1
10,000    1
312       1
789       1
1,195     1
422       1
21        1
765       1
3,500     1
945       1
326       1
3,100     1
3,390     1
1,356     1
2,599     1
472       1
454       1
328       1
396       1
291       1
Name: count, dtype: int64

Based on my investigation, I realized that the 'Price' column is the str datatype. For prices that are above 999, the value contains a column. So my solution would involved removing the comma with a an empty string. Then I can cleanly convert the 'Price' column from the str datatype to the int datatype.

In [12]:
# Replace commas in the 'Price' column with an empty string
data["Price"] = data["Price"].str.replace(',', '')

# Convert the 'Price' column to int
data["Price"] = pd.to_numeric(data["Price"])

#Results
print(f"Data type: {data['Price'].dtype}")
print(f"Missing values: {data['Price'].isna().sum()}")
print(f"\nFirst 20 values:")
print(data["Price"].head(20))
print(data["Price"].describe())

Data type: int64
Missing values: 0

First 20 values:
0     145
1      37
2      28
3     199
4     549
5     149
6     250
7      90
8     270
9     290
10    170
11     59
12     49
13     68
14    285
15     75
16    145
17    100
18    150
19    700
Name: Price, dtype: int64
count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64


### Question 1 Part 2

In [15]:
data2 = pd.read_csv("data/mn_police_use_of_force.csv")
print(f"Missing values in 'subject_injury': {data2['subject_injury'].isna().sum()}")
print(f"Proportion of missing values in 'subject_injury': {(data2['subject_injury'].isna().sum() / len(data2))*100}%")
print(data2['subject_injury'].value_counts())

data2[['subject_injury', 'force_type']].head()

Missing values in 'subject_injury': 9848
Proportion of missing values in 'subject_injury': 76.19342359767892%
subject_injury
Yes    1631
No     1446
Name: count, dtype: int64


,subject_injury,force_type
0,NaN,Bodily Force
1,NaN,Chemical Irritant
2,NaN,Chemical Irritant
3,NaN,Chemical Irritant
4,NaN,Chemical Irritant


Based on my initial look at the 'subject_injury' and the 'force_type' column, I realized that the the 'subject_injury column has many missing values. Approxiametely 76 percent of this column are missing values. Since the goal is to find the meaning behind this missing values, cross tabulating does not work at this moment because the missing value is not a unique value that is counted. So the solution I would use is to fill the missing value. 

In [ ]:
#fill missing values in the 'subject_injury' column with 'Missing'
data2['subject_injury'] = data2['subject_injury'].fillna('Missing')
print(data2[['subject_injury', 'force_type']].head())

#Perform the crosstab analysis
crosstab = pd.crosstab(data2['subject_injury'], data2['force_type'])
print(crosstab)

crosstab_analysis = pd.DataFrame(
    index=crosstab.columns,
    columns=['Frequency_Missing']
)
crosstab_analysis['Frequency_Missing'] = (crosstab.loc['Missing'].values)
print(crosstab_analysis)


After filling the missing values with the string 'Missing', I am able to see the relationship between the missing values in 'subject_injury" and force type using the cross tab. I am noticing that bodily force, chemical irritant, and maximal restraint technique of most, if not all, values as missing, which is an interesting observation. 

### Question 2
Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

In [ ]:
#Load the data
shark_df = pd.read_csv("data/GSAF5.csv", encoding="latin-1")

#Remove Empty Columns
shark_df = shark_df.dropna(axis=1, how='all')
shark_df = shark_df.drop(columns = ['Unnamed: 21', 'Unnamed: 22'])

#Display basic information about the dataframe
print(shark_df.head())
print(shark_df.info())
print(shark_df.describe())
print(shark_df.columns)

In [25]:
shark_df['Year'].describe()

count    7115.000000
mean     1936.549122
std       269.943193
min         0.000000
25%      1948.000000
50%      1987.000000
75%      2010.000000
max      2026.000000
Name: Year, dtype: float64

## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** [To be provided]
- **AI tool and model used:** [To be provided]

### *Prompts used

1. [Paste prompt]
2. [Paste follow-up prompt]

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. ...
2. ...

### *Evaluation of AI suggestions


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]